File: cpp/graph/topologicalSort.cpp
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>

using std::vector;
using std::unordered_map;
using std::unordered_set;
using std::reverse;

// Given a directed acyclical graph, return a valid
// topological ordering of the graph. 
vector<int> topologicalSort(vector<vector<int>>& edges, int n) {
    unordered_map<int, vector<int>> adj;
    for (int i = 1; i < n + 1; i++) {
        adj[i] = vector<int>();
    }
    for (vector<int> edge : edges) {
        int n1 = edge[0], n2 = edge[1];
        adj[n1].push_back(n2);
    }
    vector<int> topSort;
    unordered_set<int> visit;
    for (int i = 1; i < n + 1; i++) {
        dfs(i, adj, visit, topSort);
    }
    reverse(topSort.begin(), topSort.end());
    return topSort;
}

void dfs(int src, unordered_map<int, vector<int>>& adj,
    unordered_set<int>& visit, vector<int>& topSort) {
    if (visit.count(src) > 0) {
        return;
    }
    visit.insert(src);

    for (int neighbor : adj[src]) {
        dfs(neighbor, adj, visit, topSort);
    }
    topSort.push_back(src);
}

File: python/graph/topologicalSort.py
# Course schedule ii is easier than normal topological sort
# Graph not necessarily connected

# Given a directed acyclical graph, return a valid
# topological ordering of the graph. 
def topologicalSort(edges, n):
    adj = {}
    for i in range(1, n + 1):
        adj[i] = []
    for src, dst in edges:
        adj[src].append(dst)

    topSort = []
    visit = set()
    for i in range(1, n + 1):
        dfs(i, adj, visit, topSort)
    topSort.reverse()
    return topSort

def dfs(src, adj, visit, topSort):
    if src in visit:
        return True
    visit.add(src)
    
    for neighbor in adj[src]:
        dfs(neighbor, adj, visit, topSort)
    topSort.append(src)

edges = [[1,2],[1,3],[2,4],[3,4],[5,6],[6,4],[6,8],[7,8]]
print(topologicalSort(edges, 8))